### PR TITLE
Nh/fixes 3732 insertOrUpdate using other Realm

### LINKED ISSUE #3732: StackOverflow when using insertOrUpdate to insert object with two-way relation
#### Goal

I want to copy an object that has a two way relation from one Realm to another using insertOrUpdate. 

#### Expected Results

Item is inserted to the new realm and relations are maintained.

#### Actual Results

Stack Overflow Error
```
 at io.realm.ParentRealmProxy.insertOrUpdate(ParentRealmProxy.java:534)
                      at io.realm.ChildRealmProxy.insertOrUpdate(ChildRealmProxy.java:506)
                      at io.realm.ParentRealmProxy.insertOrUpdate(ParentRealmProxy.java:534)
                      at io.realm.ChildRealmProxy.insertOrUpdate(ChildRealmProxy.java:506)
                      at io.realm.ParentRealmProxy.insertOrUpdate(ParentRealmProxy.java:534)
                      at io.realm.ChildRealmProxy.insertOrUpdate(ChildRealmProxy.java:506)
                      at io.realm.ParentRealmProxy.insertOrUpdate(ParentRealmProxy.java:534)
                      at io.realm.ChildRealmProxy.insertOrUpdate(ChildRealmProxy.java:506)
                      at io.realm.ParentRealmProxy.insertOrUpdate(ParentRealmProxy.java:534)
```

#### Steps & Code to Reproduce

See code sample

#### Code Sample
```
public class Parent extends RealmObject {
    @PrimaryKey
    @Required
    private String id;

    private RealmList<Child> childList;

    public RealmList<Child> getChildList() {
        return childList;
    }
}

public class Child extends RealmObject {

    @PrimaryKey
    @Required
    private String id;

    private Parent parent;

    public Parent getParent() {
        return parent;
    }

    public void setParent(Parent parent) {
        this.parent = parent;
    }
}

public class RealmOverflowBugActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_realm_overflow_bug);

        Realm.init(this);

        RealmConfiguration config1 = new RealmConfiguration.Builder().name("config1").deleteRealmIfMigrationNeeded().build();
        RealmConfiguration config2 = new RealmConfiguration.Builder().name("config2").deleteRealmIfMigrationNeeded().build();

        Realm realm1 = Realm.getInstance(config1);
        Realm realm2 = Realm.getInstance(config2);

        //Set up initial data to realm1
        realm1.beginTransaction();

        //Add data
        Parent parent = realm1.createObject(Parent.class, "parentKey");
        Child child = realm1.createObject(Child.class, "childKey");

        //Set relations
        parent.getChildList().add(child);
        child.setParent(parent);

        realm1.commitTransaction();


        //Insert object with relations from realm1 to realm2
        realm2.beginTransaction();
        realm2.insertOrUpdate(parent); //StackOverflowError! Using copyToRealmOrUpdate works.
        realm2.commitTransaction();

        realm1.close();
        realm2.close();
    }
}
```

#### Version of Realm and tooling
Realm version(s): 2.1.1

Realm sync feature enabled: yes/no

Android Studio version: 2.2.1

Which Android version and device: 
Nexus5 emulator

### PR DESCRIPTION
**Pull request by nhachicha**
fixes #3732
Using an `IdentitifyHashMap` in case of a cyclic dependencies between existing Realm models, will end up in an infinite loop (_StackOverFlow_), because the cache will not return previously inserted elements (since the Identity/Hash of the newly inserted objects doesn't match the graph object passed as an argument, because traversing the graph will create a different *RealmProxy class, which makes the _cache_ useless) 

The benchmark still looks good on batch inserts
benchmark_batchwrite

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java [hunk 1]
```diff
@@ -293,6 +294,39 @@ public void insertOrUpdate_cyclicType() {
...
+        realm2.commitTransaction();
+
+        assertEquals(1, realm1.where(Owner.class).count());
+        assertEquals(1, realm1.where(Owner.class).findFirst().getDogs().size());
+        assertEquals(1, realm1.where(Dog.class).count());
+
+        assertEquals(realm1.where(Owner.class).count(), realm2.where(Owner.class).count());
+        assertEquals(realm1.where(Dog.class).count(), realm2.where(Dog.class).count());
+
+        assertEquals(1, realm2.where(Owner.class).findFirst().getDogs().size());
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java]: Is it worth testing that `realm2.where(Owner.class).findFirst().getDogs().first()` is the same as `realm2.where(Owner.class).findFirst()`?

**Comment by nhachicha** [comment]
@cmelchior yes that the issue, `copy*` methods already uses a `HashMap` so no issues.

**Comment by nhachicha** [comment]
@realm/java PR updated

**Pull request merged**